[MNGSITE-550] What's new in Maven 4? (#598)
apache/maven-site: content/markdown/whatsnewinmaven4.md

```

Maven 4 introduces the possibility to only use profiles when they exist.
To do so, the `?` argument was added to the profile parameter.
When using this, the build won't break; instead, an information message will be printed twice (at the start and end).
See the following snippet for an example:

```
> mvn compile -P?nonexistent
[INFO] The requested optional profiles [nonexistent] could not be activated or deactivated because they do not exist.
[...]
[INFO] BUILD SUCCESS
[INFO] ----------------------------------------------------------------------------------------------------------------
[INFO] Total time:  0.746 s
[INFO] Finished at: 2024-12-14T13:24:15+01:00
[INFO] ----------------------------------------------------------------------------------------------------------------
[INFO] The requested optional profiles [nonexistent] could not be activated or deactivated because they do not exist.
```

Maven 4 also introduces more flexible ways to activate profiles by providing condition-based activation.
See [MNG-8286][27] for more information about supported functions.

### Lifecycle changes

#### Lifecycle changed from graph to tree

> Up to Maven 3, the lifecycle was a mathematical graph.

Reviewer: The lifefcycle was really more a list than a graph.

Author: Rephrased it.

and needs to be updated to use JSR-330 - see [Maven & JSR-330][26] for further information.

**Advice**: If you are maintaining a Maven plugin, you should test it with Maven 3.9.x, pay close attention to upcoming
warnings, and update the plugin accordingly.

### Improved encryption

Security is important, and storing unencrypted passwords is bad practice.
Maven 3's password encryption had several serious issues and could more accurately be called "password obfuscation".
Maven 4 instead has a completely redone encryption system, based on Maven Encryption (`mvnenc`) - a standalone CLI tool.
As of now, it provides functionally equivalent operations as Maven 3 (see ["Maven: Password Encryption"][30]) but adds
improvements like a `decrypt` functionality.
A broad overview of the problems in Maven 3 and the solution in Maven 4 can be found in the
article ["Handling sensitive data in Maven"][31] by Maven maintainer Tamás Cservenák.

### Maven Resolver

The [Maven Artifact Resolver][28] is a library for working with artifact repositories and dependency resolution.
Maven 4 will include the new 2.0 release of this library, which contains more than 150 fixes, upgrades, and improvements,

> e.g., a Java native HTTP-Client - thanks to raising the Java version requirement to JDK 17!

Reviewer: I think we should mention that the resolver aims at being hidden behind the Maven 4 API and that it should not be used directly anymore.

Author: Added. I would like to write another article about plugin changes, but I will need help with that in terms of "you tell me what to write" as I have almost no clue about plugin development yet.

necessary to use an artifact as a dependency.
This POM will only keep what is really needed to use the artifact, e.g., dependency information.

**Note**: See below for a comparison of the content of both POMs.

### Model version 4.1.0

With two types of POM, Maven 4 can make additions to the Build-POM as it is only be used by Maven.
Therefore, with Maven 4, a new Model version 4.1.0 is introduced.
This version adds some new elements and attributes, while others are marked as deprecated.
To not break the ecosystem, this version is only available for the Build-POM, while the Consumer-POM will still use
version 4.0.0.
Maven generates the Consumer-POM during the build from the build POM.

**Note**: Maven 4 will continue to build your model version 4.0.0 project.
There is no need to update your POMs to 4.1.0 as long as you don't want to make use of the new features.

### Modules are now subprojects

From the early days of Maven 1 until today, all build information is stored in the POM, short for "Project Object
Model".
Together with build folders and other files, the wording "Maven project" is used.

> However, for projects containing multiple parts, e.g., an API and a client, each of those parts was called a "module"

Author: I rephrased the section and added example. However I left out a `groupId` in this, because it's not needed. Maven will build subprojects as long as they are listed as such (and existent ofc), but a subproject does not need to have the same `groupId` or have the higher project configured as parent (verified with Maven 4 RC2)

Model".
Together with build folders and other files, the wording "Maven project" is used.
However, for projects containing multiple parts, e.g., an API and a client, each of those parts was called a "module"
and listed in the `<modules>` section of the POM, leading to the terms "multi-module project".
This wording introduced some inconsistency, especially as projects without any `<modules>` section are often called "
single-module".
Since the introduction of the [Java Platform Module System][3] in Java 9, the term "module" has raised additional
confusion.

Maven 4 gets rid of this by naming "modules" as what they are - subprojects.
Model version 4.1.0 contains a new `<subprojects>` element analogous to the now deprecated, but still usable,
`<modules>` element.

**Note**: Use the terms `multi-project setup` and `single-project setup` to differentiate between a Maven project with
or without subprojects.

### New packaging type: bom

Maven 4 introduces a dedicated packaging type to provide a [Bill of Materials BOM][4] called "bom" to differentiate more
precisely between "parent POMs" and dependency-managing BOMs.

> While the new type is only available with Model Version 4.1.0, the final outcome is a full Maven 3 compatible (model

Reviewer 2: Try to be more precise. "final outcome" is unclear. Perhaps what's meant is something like "The build pom is only available in Model version 4.1 and later but the consumer pom is..."

Author: Rephrased

While the new type is only available with model Version 4.1.0 the final outcome is a full Maven 3 compatible (model
4.0.0) POM file!
For an example see the link above or
the [live coding by Maven maintainer Karl Heinz Marbaise at IntelliJ IDEA Conf 2004][5].

**Note**: With Maven 4 it's also possible to exclude dependencies which are declared by BOMs using the existing
`<exclusions>` element.

> Also note that in Maven4 importing BOMs with classifier is now possible.

Reviewer 3: ```suggestion
Also note that in Maven 4 importing BOMs with classifier is now possible.
```

`<bomClassifier>` element.
This means that an imported BOM must **not** come from the same reactor as the current build but be available outside

> the project before the build (in other words: you should import only external BOMs) or it may break your build (as

Reviewer 2: delete " or it may break your build (as shown in [MNG-8009][32])"

> Therefore, Maven team suggests that projects BOMs be generated as classified artifacts (see Maven build "skinny" vs "

Reviewer 3: ```suggestion
Therefore, the Maven team suggests that project BOMs should be generated as classified artifacts (see Maven build "skinny" vs "
```

> Therefore, Maven team suggests that projects BOMs be generated as classified artifacts (see Maven build "skinny" vs "

Reviewer 3: Apart from the grammar, I'm not sure what this means. I hope somebody else can clarify?

Author: @cstamas Can you please add a comment about this. I introduced this sentence based on your feedback on my proposal (https://github.com/Bukama/MavenStuff/pull/1/files).

Info: The terms "skinny" and "fat" BOMs are not used in the  Maven dependency guide (yet)  https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html
I only know the terms in combination with skinny WARs etc.

Reviewer 4: Our stance (@gnodet could chime in when online) is following: BOMs should be _generated_ and not used within reactor (see issues), because users are "tempted" to create BOMs within reactor (is fine, as long BOM is "published only"), and then reuse BOM in same way as any external BOM would be reused: by importing. But, importing causes "build peculiarities" usually unexpected to users. Hence, we think that _generating BOMs_ simply prevents the two issues, as reactor dependendencies should be managed using inheritance, as it was always the case with Maven. Appearance of BOMs simply caused this "boost of creativity", to reuse them within reactor, but also publish them to downstream users, and it is wrong. Hence, Maven 4 shows an "example", and generates it's own BOMs. Related https://issues.apache.org/jira/browse/MNG-8012 and https://issues.apache.org/jira/browse/MNG-8009

As per "what is a BOM", there is a split among us as can be seen in this old PR: https://github.com/apache/maven/pull/689

In short, for some, "[BOM represents only the set of modules of a multimodule](https://github.com/apache/maven/pull/689#issuecomment-1061640064)" (only the reactor, so to say) while for others (@gnodet and me included) "BOM is bill of materials", so full stack (if I do an extension for Maven version X, I want same version of resolver, logging, Sisu etc as well... Similarly, look at all our plugins:  after version property defining Maven version that plugins builds against, property is usually followed by a section with bunch of "keep in lockstep with Maven" version properties for logging, resolver, etc). Hence, the "skinny" (first definition) vs "fat" (second definition). In fact, Maven4 generates both now, so pleases both parties.

Reviewer 4: https://github.com/apache/maven/blob/575ad3719c79b061c78d531ec9d29793f159c2ec/apache-maven/pom.xml#L235

Reviewer 4: So to recap:
* BOM import should happen only with POMs not coming from reactor (in other words "you should import only external BOMs").
* BOM creation does not matter, doing it by hand is fine as well, but generating BOMs is much more less error prone, as you "just do the usual maven stuff" and generator makes you a BOM out of it. Less typos, less room for mistakes. And, it does not becomes "tempting" to import it either :smile:

Author: Thank you for the additional info here and on Slack. I rephrased the section. Can you please check if it's a) right and b) understandable?

(I have to go for some hours, but will be back later / tomorrow morning) if anything else is needed to change here.

Reviewer 4: I'd leave that to @mthmulders but am free to answer  to any new upcoming question! (and my grammar is bad as well).

Reviewer 3: The clarification helps me a lot, thanks @Bukama and @cstamas!

of the root folder in your POM files, e.g., when you want to define the path to a `checkstyle-suppressions.xml` file for
the checkstyle plugin.
Maven 4 now provides official properties to reference the root directory in your POM configuration.
The following table shows the official properties.

| Property                   |  Scope  | Definition                                                              | Always |
|:---------------------------|:-------:|:------------------------------------------------------------------------|:------:|
| `${project.rootDirectory}` | Project | `.mvn` folder or `root` attribute in pom                                |   No   |
| `${session.topDirectory}`  | Session | Current directory or `--file` argument                                  |  Yes   |
| `${session.rootDirectory}` | Session | `.mvn` folder or `root` attribute in pom for the `topDirectory` project |   No   |

As you can see, these properties differentiate by their scope, where `project` is always related to the Maven project's
definition (you could interpret this as the POM files) and `session` is the actual execution of a Maven build and is

> therefore related to the folder from where you start Maven.

Reviewer 2: folder --> directory

related to or is?

Major changes regarding plugins contain a now proper immutable plugin model together with a changed plugin API.
The updated API provides hints as a preparation for Maven 4.
You can enable them passing the following argument to your build: `-Dmaven.plugin.validation=verbose`.
You should also only rely on the official Maven BOMs when developing plugins.
If a plugin still relies on long time deprecated and now removed Plexus dependency resolution, it will no longer work
and needs to be updated to JSR-330 - see [Maven & JSR-330][26] for further information.

**Advice**: If you are maintaining a Maven plugin, you should test it with Maven 3.9.x, have a close look at upcoming
warnings and update the plugin.

### Improved encryption

Security is important and storing unencrypted passwords is bad.
Maven 3's password encryption has several serious issues and could rather be call a “password obfuscation”.
Maven 4 instead got a fully redone encryption, based on Maven Encryption (`mvnenc`) - a standalone CLI tool.
As for now it provides functionally equivalent operations as Maven 3 (see ["Maven: Password Encryption"][30]), but adds
improvements like a `decrypt` functionality.

> A brought overview about the problems in Maven 3 and the solution in Maven 4 can be found in the

Reviewer 3: ```suggestion
A broad overview about the problems in Maven 3 and the solution in Maven 4 can be found in the
```

article ["Handling sensitive data in Maven"][31] by Maven maintainer Tamás Cservenák.

### Maven Resolver

The [Maven Artifact Resolver][28] is a library for working with artifact repositories and dependency resolution.
Maven 4 will include the new 2.0 release of this library, which contains more than 150 fixes, upgrades and improvements,
e.g. a Java native HTTP-Client - thanks to lift of the Java version to JDK 17!

> See the project page for more information!

Reviewer 3: I would suggest to make this a link, or remove this line. If you go for a link, I would suggest to remove the exclamation mark.

Author: Hmm, line 381 contains the link to the project page. But yeah sounds a bit weired. Will remove it.

### Maven Shell

When you run any `mvn` the full chain of booting Java, Maven, load config, do the work, tear down and exit the process
is done - **every single time**.
To improve performance and reduce build time you can use the [Maven Daemon][29] (`mvnd`), which manages a pool of N
resident Maven processes.
With Maven 4 you can also use the brightly polished "Maven Shell" (`mvnsh`), which will keep exactly one resident Maven
process as long as the shell is opened.

> Stay tuned for more information about this!

Reviewer 3: If there is no more information available, I would not mention this. We can always update the article when there is a more detailed blog/video/resource available.

```suggestion
Stay tuned for more information about this!
```

Author: Removed it. The JIRA-Issue and the PR about the shell do not contain any docs (https://github.com/apache/maven/pull/1982/files). I think @cstamas will write a blog post about it (as 2nd part of https://cstamas.org/blog/2024/12/maven4-ante-portas/ I think?) or add docs later to the official site.

Reviewer 3: I'm sorry, I meant to suggest only to keep the last sentence out. I do appreciate that you mentioned the Maven shell - it's just that I found the "stay tuned" a little less helpful.

[...]
[INFO] BUILD SUCCESS
[INFO] ----------------------------------------------------------------------------------------------------------------
[INFO] Total time:  0.746 s
[INFO] Finished at: 2024-12-14T13:24:15+01:00
[INFO] ----------------------------------------------------------------------------------------------------------------
[INFO] The requested optional profiles [nonexistent] could not be activated or deactivated because they do not exist.
```

Maven 4 also introduces more flexible ways to activate profiles by providing condition-based activation.
See [MNG-8286][27] for more information about supported functions.

### Lifecycle changes

#### Lifecycle changed from graph to tree

In Maven 3 the lifecycle is an ordered list containing all phases.
This changed with Maven 4, where the lifecycle is defined as a tree of phases.
This allows for more consistent execution of dependent phases.
For example, the `compile` phase must execute after `compile-only` project dependencies have reached the `ready` phase.
Maven 4 also allows to skip phases (in comparison to the old graph).
For example, it's possible to `deploy` an artifact without `install`ing it to the user repository.

> 

Reviewer: consistent -> fine grained

The problem is not so much about consistency, rather than expressing dependencies closer to the reality. 
For compatibility, the behavior is not really changed by default compared to Maven 3.  In order to actually leverage the new tree-based lifecycle, the user has to enable the `concurrent` builder using `-b concurrent`.  Running with the concurrent builder will allow finer grained dependency instead of the usual Maven 3 per-project dependency.  With the usual builders and run `mvn verify`, a project will run up to the verify phase, before a dependent project can start being built, whereas the concurrent builder will allow a dependant project to be built as soon as the dependencies are at the `ready` phase.  Same for `install` vs `deploy`, this is only possible when using the concurrent builder (for compatibility).

Author: Thank you for the explanation - I added it to the article for a better understanding about this feature.

#### Pre- and post-phases, ordering of executions

> Every lifecycle phase now has a `before-` and an `after-` phase, allowing plugins to bind themselves to those.

Reviewer: The prefixes are `before:` and `after:` with `:` instead of `-`.
The same applies to the whole section of course.

Each project within a subdirectories is called "subproject".

Example: A project A contains a subdirectory which contains its own POM file for project B.
We say that B is a subproject of A.

However, Maven does not build subprojects as long as they are not linked as such in the POM file.
In the past this was done by listing each subproject in the `<modules>` section of the parent's POM.
The inconsistent wording led to the situation that many people calling subprojects "modules".
Projects which have at least one subproject are often called "multi-module project", while those which don't have any
are "single-module project".
Since the introduction of the [Java Platform Module System][3] in Java 9, the term "module" has raised additional
confusion inside the Maven community.

Maven 4 gets rid of this by renaming modules to projects and submodules to subprojects.
Model version 4.1.0 contains a new `<subprojects>` element analogous to the now deprecated, but still usable,
`<modules>` element.

**Note**: Use the terms `multi-project setup` and `single-project setup` to differentiate between a Maven project with
or without subprojects.

### New packaging type: bom

> Maven 4 introduces a dedicated packaging type to provide a [Bill of Materials BOM][4] called "bom" to differentiate more

Reviewer 2: Maven 4 introduces a dedicated `bom` packaging type to provide a [Bill of Materials BOM][4]. This now differentiates between parent POMs and dependency-managing BOMs.

precisely between "parent POMs" and dependency-managing BOMs.
The new type is only available as a Build-POM in Model Version 4.1.0 and later, but Maven generates a full Maven 3
compatible Consumer-POM during the build.
For an example, see the link above or
the [live coding by Maven maintainer Karl Heinz Marbaise at IntelliJ IDEA Conf 2024][5].

**Note**: With Maven 4, it's also possible to exclude dependencies that are declared by BOMs using the existing
`<exclusions>` element.
Also note that in Maven 4, importing BOMs with a classifier is now possible.
Therefore, the Maven team suggests that project BOMs should be generated as classified artifacts, using the
`<bomClassifier>` element.
This means that an imported BOM must **not** come from the same reactor as the current build but be available outside

> the project before the build (in other words: you should import only external BOMs).

Reviewer 2: (in other words: you should import only external BOMs).
-->
. In other words: you should only import external BOMs.

### Comparing Build-POM and Consumer-POM

The following table shows a rough comparison of which content is available in which POM type when using Maven 4.

**Notes**:

* The column "Consumer-POM" does not apply to artifacts that are of type "pom", because "pom"-artifacts are designed to
  contain build information, e.g. plugin configuration!
* Some of the build-related content which is (as of now) still available in the Consumer-POM might be available only in
  the Build-POM in the future.

| Content                                    | Build-POM | Consumer-POM |
|:-------------------------------------------|:---------:|:------------:|
| Model version                              |   4.1.0   |    4.0.0     |
| 3rd party dependency information           |     ✅     |      ✅       |
| POM properties                             |     ✅     |      ❌       |
| Plugin configuration                       |     ✅     |      ❌       |
| Repository information                     |     ✅     |      ✅       |
| Project information / environment settings |     ✅     |      ✅       |
| Deployment to remote repository            |     ✅     |      ✅       |

> **Warning**: There are rare situations where Maven 4 will produce a Consumer-POM based on version 4.1.0, e.g., when

Reviewer 2: This feels like a bug we should fix before releasing 4.0 (not in this PR)

condition-based profiles (see below) can't be transformed to version 4.0.0.
Maven will show a warning in such situations.

### Declaring the root directory and directory properties

> Every time Maven executes a build, it has to determine the project's root to identify things like the parent project,

Reviewer 2: The language here is confusing. I think the root folder is where the pom.xml resides. This is something else. One or the other must be renamed.

Author: As written yesterday, in the PR/issue. @gnodet wanted to request changes anyway as he said a lot in this article is wrong, so he can clarify this section too (He is the author of the PR which introduced this).

Author: Created an issue to describe the terms in a separate PR.

of the root folder in your POM files, e.g., when you want to define the path to a `checkstyle-suppressions.xml` file for
the checkstyle plugin.
Maven 4 now provides official properties to reference the root directory in your POM configuration.
The following table shows the official properties.

| Property                   |  Scope  | Definition                                                              | Always |
|:---------------------------|:-------:|:------------------------------------------------------------------------|:------:|
| `${project.rootDirectory}` | Project | `.mvn` folder or `root` attribute in pom                                |   No   |
| `${session.topDirectory}`  | Session | Current directory or `--file` argument                                  |  Yes   |
| `${session.rootDirectory}` | Session | `.mvn` folder or `root` attribute in pom for the `topDirectory` project |   No   |

As you can see, these properties differentiate by their scope, where `project` is always related to the Maven project's
definition (you could interpret this as the POM files) and `session` is the actual execution of a Maven build and is the

> directory from where you start Maven.

Reviewer 2: sounds like the current working directory, and we should probably just say that

As a consequence of the definition, it's clear that the `rootDirectory` can only contain a value when either a `.mvn`
directory is defined or the `root` attribute is set to true.

> However, if defined, it should always have the same value for a given project, whereas the value of the `topDirectory`

Reviewer 2: Maybe one is a root directory and one is a top directory? This needs to be majorly clarified before we release 4.0. It might require code changes.

property can change depending on the execution point.

> Keep in mind that the root directory of the whole project (when considering multiple subprojects) is different from each

Reviewer 2: OK, so we now have top, root, and base directories

> subproject's own base directory, which is still accessible via the `${basedir}` property for use in POM

Reviewer 2: delete still

files and will always have a value.

**Note:** In the past, some people "hacked" workarounds for the `rootDirectory` properties, mostly by using internal
properties.

> Starting with Maven 4, those "hacks" will most probably not work anymore because some of those properties were removed

Reviewer 2: will they work or not? Be definite about this.

> or at least marked as deprecated.

Reviewer 2: delete "at least"

  <dependencies>
    <dependency>
      <groupId>the.dependent.subproject.groupId</groupId>
      <artifactId>the.dependent.subproject.artifactId</artifactId>
    </dependency>
  </dependencies>
</project>
```

### Full support of CI-friendly variables

Maven 3.5.0 introduced partial support for CI-friendly variables, e.g., `${revision}`, in your POM files.
However, this still required the usage of the [Flatten Maven Plugin][20] for full functionality.
Since Maven 4, no additional plugin is needed; full built-in support is provided.
You can now use variables as versions in your configuration, e.g.,

```xml

<groupId>my.groupId</groupId>
<artifactId>my.artifactId</artifactId>
<version>${revision}</version>
```

You have to provide a value for this variable when starting the build, for example by using a `maven.config` file.
In CI pipelines it's commonly done using a parameter, e.g., `mvn verify -Drevision=4.0.1`.

Maven maintainer Karl Heinz Marbaise shows a larger example in
his [article "Maven 4 - Part I - Easier Versions" (2024)][21].

### Reactor improvements and fixes

Building a project with multiple subprojects could cause trouble when one subproject depended on one of the

> others and its own build failed for whatever reason.

Reviewer 2: antecedent of its is unclear; rewrite

**Recommendation: Do not use `mvn clean install`.
Instead, use `mvn verify` for your regular builds!**

To improve usability when resuming a failed build, you can now use `--resume` or its short parameter `-r` to resume a
build from the subproject that last failed.
So you don't have to manually pass the name of the failed subproject as the starting point to resume from.
The reactor is now also aware of successfully built subprojects when the overall build failed and will skip rebuilding
those if you resume the build.  
With Maven 4, it's also aware of subfolder builds [MNG-6118][12], which becomes pretty handy when you only want to
execute tools (e.g., Jetty) on/with certain subprojects, but not on every subproject.
See Maven maintainer Maarten Mulders's article ["What's New in Maven 4" (2020)][13] for a small example.

### Further improvements

Further improvements to subprojects will also improve daily work with them.
Thanks to [MNG-6754][14], all subprojects will now have consistent timestamps in their packaged archives, while in Maven
3, each subproject had a different one.
This should make it easier to identify the archives that belong together.

> When using Maven 3, deploying a project with multiple subprojects could end up in a situation where some (successfully

Reviewer 2: some (successfully
built) --> successfully
built

> built) subprojects were deployed to the (local or remote) repository, but failed subprojects were obviously not.

Reviewer 2: delete obviously

> This was finally changed in Maven 4 to the way most users expect:

Reviewer 2: the way --> what

Only deploy when all subprojects are built successfully.

## Workflow, lifecycle and runtime changes

### Java 17 required to run Maven

> The required Java version to run Maven 4 will be Java 17.

Reviewer 2: will be --> is

This allows Maven (and its maintainers) to make use of newer language features and improvements brought by the
JDK.

> **Important note**: Java 17 will only be needed to **run Maven**!

Reviewer 2: will only be --> is only

> You will still be able to compile against older Java versions using the same [compiler plugin configuration][6] as

Reviewer 2: will still be able to --> can still

before.

> If this does not fit your requirements because you need to compile against or use another JDK, see

Reviewer 2: delete "this does not fit your requirements because"

the [Guide to Using Toolchains][7] (or the article [Introduction to Maven Toolchains][8] by Maven maintainer Maarten
Mulders).

### Application maintenance

As with every major update, extensive application maintenance has occurred, including significant code, API, and
dependency updates.
For example, the "Plexus Containers" dependency injection was removed - after being deprecated since Maven 3.2 (2010)!
Code updates include not only newer Java language features but also changes to make maintenance easier and
less time-consuming.
This also includes removing features that either should never have worked or were only kept for backward compatibility

> already in Maven 3, e.g., using `${pom.*}` expressions.

Reviewer 2: delete already

**Note**: Due to upgrading the default versions of Maven plugins, your build might behave differently than before, even
if you didn't purposely change anything.
To avoid this situation, always define fixed versions of all the plugins you use.
By doing this, you are in control of your build.
Maven 4 will issue a warning if you rely on default versions defined in Maven's Super POM.

### "Fail on severity" parameter

Maven 4 introduces a "fail on severity" build parameter, which breaks the build when the severity of at least one log
message matches the given argument.

The parameter can either be used by its full name (`--fail-on-severity`) or as a short handle (`-fos`).
The parameter is followed by an argument specifying a log level severity, e.g., `WARN`.

### Improvements to profiles

Trying to use a nonexistent profile in a build causes the build to fail, as the following command line snippet shows:

```
> mvn compile -Pnonexistent
[ERROR] The requested profiles [nonexistent] could not be activated or deactivated because they do not exist.
```

Maven 4 introduces the possibility to only use profiles when they exist.
To do so, the `?` argument was added to the profile parameter.

> When using this, the build won't break; instead, an information message will be printed twice (at the start and end).

Reviewer 2: break; instead,
-->
break. Instead,

See the following snippet for an example:

```
> mvn compile -P?nonexistent
[INFO] The requested optional profiles [nonexistent] could not be activated or deactivated because they do not exist.
[...]
[INFO] BUILD SUCCESS
[INFO] ----------------------------------------------------------------------------------------------------------------
[INFO] Total time:  0.746 s
[INFO] Finished at: 2024-12-14T13:24:15+01:00
[INFO] ----------------------------------------------------------------------------------------------------------------
[INFO] The requested optional profiles [nonexistent] could not be activated or deactivated because they do not exist.
```

Maven 4 also introduces more flexible ways to activate profiles by providing condition-based activation.
See [MNG-8286][27] for more information about supported functions.

### Lifecycle changes

#### Lifecycle changed from graph to tree

> Up to Maven 3, the lifecycle was an ordered list containing all phases.

Reviewer 2: Up to --> In Maven 3 the lifecycle is

This changed with Maven 4, where the lifecycle is defined as a tree of phases.
This allows for more consistent execution of dependent phases, e.g., `compile` must execute after `compile-only` project

> dependencies have reached the `ready` phase, and also allows "skipping" phases (in comparison to the old graph), e.g.,

Reviewer 2: run-on sentence

`deploy` an artifact without `install`ing it to the user repository.

#### Pre- and post-phases, ordering of executions

> Every lifecycle phase now has a `before` and `after` phase, allowing plugins to bind themselves to those by adding their

Reviewer 2: and an

> prefixes to the name of the main phase.

Reviewer 2: antecedent of their is unclear

For example, if you want to set up test data before running your integration tests, you could execute tasks during the
`before-integration-test` phase.

If this is not enough, perhaps because you want to do multiple things inside the same phase, you can order each

> execution inside a phase by using square brackets with an integer at the end of the phase name, e.g.,

Reviewer 2: name, For example,

(In general avoid Latin: i.e., e.g., n.b., etc.)

```
before-integration-test[100]
before-integration-test[200]
```

**Warning**: The conceptual `pre-*` and `post-*` phases, which were only available for selected phases and had

> inconsistent naming, are deprecated - don't use them!

Reviewer 2: - don't
. Don't

> This becomes even more important if you were binding a plugin to the `post-*` phase of a lifecycle phase because the

Reviewer 2: were --> are

> `pre-*` phase of the phase you really wanted to bind to did not exist, e.g., binding to `process-resources` phase

Reviewer 2: wanted --> want
did -- does

In general, prefer present tense

because there was no `pre-compile` phase.

> ## Maven plugins, security and tools

Reviewer 2: security,

Author: argh. I try to remind me about this difference between English and German, but sometimes I miss it :(

# What's New in Maven 4?

Maven is over 20 years old, and is one of the most used build tools in the Java world.
Throughout the years, one important rule has been maintaining the highest backward compatibility possible, especially

> with its [POM-schema with Model version 4.0.0][2], used not only for the build itself but also by consumers.

Reviewer 2: I'm not sure we can expect people reading this document to already know the difference between build and consumers. This should probably be carefully explained right up front.

Also, "consumers" is a strange term here. I might consider dependents instead, but either way it does need to be defined.

Reviewer 2: Ping

This made Maven more than a tool; it became a whole ecosystem with many dependencies on the POM, especially the Maven
Central repository, other build tools, and IDEs.

> But this stable schema comes at a price - the lack of flexibility.

Reviewer 2: TBH, that's not totally true. It could have been stable and much more flexible than it is. The original Maven designers made some common mistakes in XML that locked POMs down far more than they had to be. It could have been stable and flexible.  Maven 4 does not seem to have improved in this regard.

Author: > Maven 4 does not seem to have improved in this regard.

I agree on this, but I think that's not the direct, but only an indirect goal of Maven 4 as stated in the line below he quote:

> Maven 4 will prepare for changes which are impossible nowadays, like a completely new build schema.

As this is what's needed: Seperate build-pom from consumer one. So maven can do what ever it wants do to without breaking the ecosystem.

May I ask how you would phrase the introduction of the article to point out the "same POM-problem?

Reviewer 2: delete "But this stable schema comes at a price - the lack of flexibility."

Author: > delete "But this stable schema comes at a price - the lack of flexibility."

deleted

> "With the Maven build schema preserved in amber, we can’t evolve much: we’ll stay forever with Maven 3 minor releases,
> unable to implement improvements that we imagine will require seriously updating the POM schema…"
> &mdash; <cite>[Hervé Boutemy (in Javaadvent 2021)][1]</cite>

> Maven 4 will prepare for changes which are impossible nowadays, like a completely new build schema.

Reviewer 2: Ping

> Maven 4 will prepare for changes which are impossible nowadays, like a completely new build schema.

Author: Updated in next push.

> Another pain point of Maven 3 is a codebase with a lot of deprecated, convoluted, non-performant, and duplicate code

Reviewer 2: Please avoid the word "performant". It's ugly jargon. If you mean slow, say slow.

Author: Removed

> which costs the volunteers who maintain Maven a lot of time.

Reviewer 2: volunteers --> developers

Not everyone is a volunteer, at least not all the time.

Author: Changed

> This means that the Maven codebase contains not only old Java code that can be optimized nowadays but also old

Reviewer 2: I'm not aware of any code in Maven that can be optimized now that couldn't be optimized in Java 7 and earlier.

Author: > I'm not aware of any code in Maven that can be optimized now that couldn't be optimized in Java 7 and earlier.

@khmarbaise told me several times of the last year that code was optimized first with Java 8 features, like streams etc and can be / will now be (since the vote for Java 17 has passed) further optimized with other language features like records, text blocks etc.
Passive benefits of improved GC, security updates inside the JDK etc. add to this.

Reviewer 2: None of this is likely to result in noticeably faster code. 

Improved GC and security updates come from running in a new JDK, not from using a higher language level or new language features.

Author: > None of this is likely to result in noticeably faster code.
> 
> Improved GC and security updates come from running in a new JDK, not from using a higher language level or new language features.

Optimized code and faster code are (like a rectangel and square) not the same - at least not for me.

Reviewer 2: ![image](https://github.com/user-attachments/assets/3987287d-77ed-4676-b3a6-9ace9483feda)

Author: Removed the part "that can be optimized nowadays". Hope this pleases you better now.

dependencies and poor API design of its own APIs, especially for Maven plugins.

> Therefore, Maven 4 will also be a maintenance release.

Author: What I want to express with that, that the Maven codebase and API got a lot of updates and improvements. It should not be a promise, but a note about this. If we don't do any more updates - the sentence in the notes stays true - the team did put a lot of effort into updates

> Therefore, Maven 4 will also be a maintenance release.

Reviewer 2: Consider the audience of this document. This simply is not relevant to someone who simply wants to use Maven to build their code. It's only relevant to the people developing Maven and that's not who this is written for.

> Therefore, Maven 4 will also be a maintenance release.

Reviewer 2: Ping

> Therefore, Maven 4 will also be a maintenance release.

Author: Decided to drop those lines.

> This article presents and explains major changes brought by Maven 4, grouped into several topics.

Reviewer 2: Ping

> This article presents and explains major changes brought by Maven 4, grouped into several topics.

Author: From my experience and the feedback I got of writing all kind of software project related articles addressing developers, managers or customers for almost 20 years an introduction into articles or documents like this helps the reader a lot.

I accept that you have another standing about this, but as the author I won't drop the introduction, but rather accept that the PR gets declined.

> This article presents and explains major changes brought by Maven 4, grouped into several topics.

Reviewer 3: When reviewing, I appreciated these initial paragraphs. They gave me the context and an idea of what to expect going forward. From my side, +1 for keeping them.

This POM will only keep what is really needed to use the artifact, e.g., dependency information.

**Note**: See below for a comparison of the content of both POMs.

### Model version 4.1.0

With now having two types of POM, Maven 4 can already make additions to the Build-POM as it will only be used by Maven (
and of course IDEs).
Therefore, with Maven 4, a new Model version 4.1.0 is introduced.
This version introduces some new elements and attributes, while others are marked as deprecated.
To not break the ecosystem, this version is only available for the Build-POM, while the Consumer-POM will still use
version 4.0.0.
This means Maven will generate the Consumer-POM during the build.

**Note**: Maven 4 will of course continue to build your model version 4.0.0 project!
There is no need to update your POMs to 4.1.0 as long as you don't want to make use of the new features.
But as with every software update - it's suggested to update them to 4.1.0, e.g., to avoid problems when deprecated
features are removed in future updates.

### Modules are now subprojects

From the early days of Maven 1 until today, all build information is stored in the POM, short for "Project Object
Model".

> Together with build folders and other files, the wording "Maven project" is used.

Reviewer 2: Ping

> Together with build folders and other files, the wording "Maven project" is used.

Author: > This all need a better explanation with examples.

Please lay out what you mean. Shall I write/show an example of a Maven project's folder? This is done the users guide (https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html). And I don't think we need to explain this to someone who's interesting in information about the changes between Maven 3 and 4.

`<modules>`.

**Note**: It's suggested to use the terms `multi-project setup` and `single-project setup` to differentiate between a
Maven project with or without subprojects.

### New packaging type: bom

Maven 4 introduces a dedicated packaging type to provide a [Bill of Material BOM][4] called "bom" to differentiate more
precisely between "parent POMs" and dependency-managing BOMs.
While the new type is only available with Model Version 4.1.0, the final outcome is a full Maven 3 compatible (model
4.0.0) POM file!
For an example, see the link above or
the [live coding by Maven maintainer Karl Heinz Marbaise at IntelliJ IDEA Conf 2024][5].

**Note**: With Maven 4, it's also possible to exclude dependencies that are declared by BOMs using the existing
`<exclusions>` element.
Also note that in Maven 4, importing BOMs with a classifier is now possible.
Therefore, the Maven team suggests that project BOMs should be generated as classified artifacts, using the
`<bomClassifier>` element.
This means that an imported BOM must **not** come from the same reactor as the current build but be available outside

> the project before the build (in other words, "you should import only external BOMs") or it may break your build (as

Author: "may",  as it's not 100%, but chances are good - that's how I understood the linked issue and what Tamas said to me.

> shown in [MNG-8009][32]).

Author: > extraneous closing parenthesis

I don't think so, the opening is at the end of the line above (but line break at 120 chars)

> (as shown in [MNG-8009][32])

That's why Maven 4.0 will show a warning if a BOM comes from the same reactor.
In the future, this will most probably be changed to make the build fail.

### Comparing Build-POM and Consumer-POM

The following table shows a rough comparison of which content is available in which POM type when using Maven 4.

**Notes**:

> * The column "Consumer-POM" obviously does not apply to artifacts that are of type "pom"!

Author: > That's not obvious to me

Explanation added (will be visible after next push)

* Some of the build-related content which is (as of now) still available in the Consumer-POM might be available only in
  the Build-POM in the future.

| Content                                    | Build-POM | Consumer-POM |
|:-------------------------------------------|:---------:|:------------:|
| Model version                              |   4.1.0   |    4.0.0     |
| 3rd party dependency information           |     ✅     |      ✅       |

> | Properties                                 |     ✅     |      ❌       |

Author: Don't understand what you mean by this.

> | Properties                                 |     ✅     |      ❌       |

Reviewer 2: Are these Java system properties?

> | Properties                                 |     ✅     |      ❌       |

Author: > Are these Java system properties?

Hm thought it is clear that this section is about POM properties. Do you think "POM properties" is clearer or maybe "Propertiy definitions used in the POM"?

### Declaring the root directory and directory variables

Every time a Maven build is executed, it has to determine the project's root to identify things like the parent project,
directory information, and so on.
To "help" Maven find the root folder, you can create a `.mvn` folder in your root directory.
This folder is intended to contain project-specific configuration to run Maven, e.g., a `maven.config` or `jvm.config`
file, and therefore was also considered as the root folder.
With Maven 4, there is a second option to clearly define the root folder.
Model version 4.1.0, usable for the Build-POM, adds a boolean attribute called `root` in the `<project>` element.
When this attribute is set to true (default is false), the directory of this POM file is considered the root directory.

Another pain point in relation to the root directory is that until Maven 4, there was no official variable to make use
of the root folder in your POM files, e.g., when you want to define the path to a `checkstyle-suppressions.xml` file for
the checkstyle plugin.
Maven 4 now provides official variables to reference the root directory in your POM configuration.

> The following table shows the official variables.

Author: > variables or properties?

Good catch! Properties is the correct term, updating.

| `${session.rootDirectory}` | Session | `.mvn` folder or `root` attribute in pom for the `topDirectory` project |   No   |

As you can see, these variables differentiate by their scope, where `project` is always related to the Maven project's
definition (you could interpret this as the POM files) and `session` is the actual execution of a Maven build and is
therefore related to the folder from where you start Maven.
As a consequence of the definition, it's clear that the `rootDirectory` can only contain a value when either a `.mvn`
folder is defined or the `root` attribute is set to true.
However, if defined, it should always have the same value for a given project, whereas the value of the `topDirectory`
variable can change depending on the execution point.

Keep in mind that the root directory of the whole project (when considering multiple subprojects) is different from each
subproject's own base directory, which was and is still accessible via the `${basedir}` property for use in POM
files and will always have a value.

**Note:** In the past, some people "hacked" workarounds for the `rootDirectory` variables, mostly by using internal
variables.

> Starting with Maven 4, those "hacks" will most probably not work anymore because some of those variables were removed or

Author: I depends :D

Some of the properties mentioned in the PR are now deprecated and some are now also "hidden"(but it's not written what "hidden" means). Others are gone as they were renamed, e.g. `executionRootDir` got renamed to `topdir`. So depending on which variables a "hacker" is using nowadays, it will or won't work in the future. I think we can't write a more specific information / warning here.

at least marked as deprecated.
See JIRA issue [MNG-7038][15] and the related [Pull Request for MNG-7038][16] for more information.

### Alternate POM syntaxes

While the syntax for the 4.0.0 Consumer-POM is set in stone for accessing the central repository, the Build-POM should
be able to evolve.
This includes allowing the use of alternate syntaxes by having Maven 4 provide a ModelParser SPI ([MNG-7836][24]),
which can be implemented as a core extension and allow the usage of a different file as the POM and allow a custom
syntax.

One of the first projects that uses this feature is the [Apache Maven Hocon Extension][25].

## Improvements for subprojects

### Automatic versioning

Maven 4 finally ships one of the oldest improvement requests - automatic parent versioning ([MNG-624][17], created in
July 2005 and originally planned for Maven 2)!
As expected, it's no longer required to define the parent versions in each subproject when using the new model version
4.1.0.
This is also extended to dependencies of project own subprojects and reduces the need to update POM files for new

> versions even more!

Author: > exclamation point

Changed (Sorry I'm so excited about automatic versioning 😍 )

those if you resume the build.  
With Maven 4, it's also aware of subfolder builds [MNG-6118][12], which becomes pretty handy when you only want to
execute tools (e.g., Jetty) on/with certain subprojects, but not on every subproject.
See Maven maintainer Maarten Mulders's article ["What's New in Maven 4" (2020)][13] for a small example.

### Further improvements

Further improvements to subprojects will also improve daily work with them.
Thanks to [MNG-6754][14], all subprojects will now have consistent timestamps in their packaged archives, while in Maven
3, each subproject had a different one.
This should make it easier to identify the archives that belong together.
When using Maven 3, deploying a project with multiple subprojects could end up in a situation where some (successfully
built) subprojects were deployed to the (local or remote) repository, but failed subprojects were obviously not.
This was finally changed in Maven 4 to the way most users expect:
Only deploy when all subprojects are built successfully.

## Workflow, lifecycle and runtime changes

### Java 17 required to run Maven

> The required Java version to run Maven 4 will be Java 17.

Reviewer 2: This should be up top

Author: > This should be up top

I understand what you mean.

Let me show you my thoughts why this changes is mentioned "so late": I decided to start with POM changes, because I think the introduction of the Build-POM/Consumer-POM is the biggest motivation for Maven 4 (as written in the introduction) and the most meaningful change of all. 
POM changes also effect the renamed subprojects, which lead to the chapter about those. And then workflow, lifecyle, runtime with Java 17 as the most important in this category. 
I was thinking about pulling the Java 17 change out of this, but decided against it, so this whole chapter reads a bit more well-rounded (not sure if this is the correct English word). I think the current structure contains a good "central thread / dramatic line" (again not sure if this is a proper translation, closest I found in the dictionary).

Reviewer 2: The whole build consumer POM distinction can be ignored by a lot of users. However their existing build just fails when they try to run with Java 11 like they did with maven 3.9.9.

This not only allows Maven (and its maintainers) to make use of newer language features and improvements brought by the

> JDK but also comes with a more secure runtime as Java 17 includes more security features than earlier versions.

Reviewer 2: I don't believe this is true

Author: According to what [Oracles writes about Security updates](https://docs.oracle.com/en/java/javase/22/migrate/security-updates.html#GUID-9C6409FA-3EE2-4717-A1B4-B9A01690C081) I think this sentences is not wrong.
But yeah it's a little bit of "manager buzzword game" 👼 

(Ana Maria Mihalcean also has a talk [From Java 17 to 21: A Showcase of JDK Security Enhancement] about this, but focusing more on security features between 17 and 21, which do not apply to Maven 4 when considering using a JDK 17 only).

Reviewer 2: You can run Maven 3 in Java 17 and get all those updates. None of them are specific to Maven and Maven 4 doesn't enable any of them compared to Maven 3.

Author: Removed that part about security.

**Important note**: Java 17 will only be needed to **run Maven**!
You will still be able to compile against older Java versions using the same [compiler plugin configuration][6] as

> before!

Reviewer 2: no exclamation point

> If this does not fit your requirements because you need to compile against or use another JDK, please look at

Reviewer 2: "please look at" --> see

the [Guide to Using Toolchains][7] (or the article [Introduction to Maven Toolchains][8] by Maven maintainer Maarten
Mulders).

> *Side information: The ballot about the required Java version was held in March 2024, shortly before Java 22 was

Reviewer 2: delete this note; it's not important

Author: > delete this note; it's not important

Hm okay. I wanted to inform, why not Java 21.

released. One reason Java 17 was chosen over Java 21 was that it was (at this time) the second-last Java version for
which many vendors offer long-term support.*

### Application maintenance

As with every major update, extensive application maintenance has occurred, including significant code, API, and

> dependency updates and even removals.

Reviewer 2: delete and even removals.

For example, the "Plexus Containers" dependency injection was removed - after being deprecated since Maven 3.2 (2010)!

> Code updates include not only the use of newer Java language features but also changes to make maintenance easier and

Reviewer 2: delete the use of

less time-consuming.
This also includes removing features that either should never have worked or were only kept for backward compatibility
already in Maven 3, e.g., using `${pom.*}` expressions.
Maven's own Super POM was also upgraded, which declares new default versions of Maven's core plugins.

**Note**: Due to upgrading the default versions of Maven plugins, your build might behave differently than before, even
if you didn't purposely change anything.

> To avoid this situation, you should always define fixed versions of all the plugins you use!

Reviewer 2: delete "you should" and no exclamation point

> By doing this, you are in control of your build - at the cost of being responsible for upgrading the versions yourself.

Reviewer 2: delete "- at the cost of being responsible for upgrading the versions yourself."

> Maven 4 will issue a warning if you rely on default versions defined in Maven's Super POM!

Reviewer 2: delete exclamation point

### "Fail on severity" parameter

> Maven 4 introduces a "fail on severity" build parameter, which will break the build when at least one log message

Reviewer 2: will break --> breaks

> matches the given argument.

Reviewer 2: it sounds like the message doesn't have to match, but rather the severity

```

Maven 4 introduces the possibility to only use profiles when they exist.
To do so, the `?` argument was added to the profile parameter.
When using this, the build won't break; instead, an information message will be printed twice (at the start and end).
See the following snippet for an example:

```
> mvn compile -P?nonexistent
[INFO] The requested optional profiles [nonexistent] could not be activated or deactivated because they do not exist.
[...]
[INFO] BUILD SUCCESS
[INFO] ----------------------------------------------------------------------------------------------------------------
[INFO] Total time:  0.746 s
[INFO] Finished at: 2024-12-14T13:24:15+01:00
[INFO] ----------------------------------------------------------------------------------------------------------------
[INFO] The requested optional profiles [nonexistent] could not be activated or deactivated because they do not exist.
```

Maven 4 also introduces more flexible ways to activate profiles by providing condition-based activation.
See [MNG-8286][27] for more information about supported functions.

### Lifecycle changes

#### Lifecycle changed from graph to tree

> Up to Maven 3, the lifecycle was literally an ordered list containing all phases.

Reviewer 2: delete literally

This changed with Maven 4, where the lifecycle is defined as a tree of phases.
This allows for more consistent execution of dependent phases, e.g., `compile` must execute after `compile-only` project
dependencies have reached the `ready` phase, and also allows "skipping" phases (in comparison to the old graph), e.g.,
`deploy` an artifact without `install`ing it to the user repository.

#### Pre- and post-phases, ordering of executions

Every lifecycle phase now has a `before` and `after` phase, allowing plugins to bind themselves to those by adding their
prefixes to the name of the main phase.
For example, if you want to set up test data before running your integration tests, you could execute tasks during the
`before-integration-test` phase.

If this is not enough, perhaps because you want to do multiple things inside the same phase, you can order each
execution inside a phase by using square brackets with an integer at the end of the phase name, e.g.,

```
before-integration-test[100]
before-integration-test[200]
```

**Warning**: The conceptual `pre-*` and `post-*` phases, which were only available for selected phases and had

> inconsistent naming, are deprecated - don't use them (anymore)!

Reviewer 2: delete (anymore)!

This becomes even more important if you were binding a plugin to the `post-*` phase of a lifecycle phase because the
`pre-*` phase of the phase you really wanted to bind to did not exist, e.g., binding to `process-resources` phase
because there was no `pre-compile` phase.

## Maven plugins, security and tools

### Maven plugins

> As mentioned above, Maven 4 will contain significant code and API updates, resulting in breaking changes for (very) old

Reviewer 2: will contain --> contains

> Maven plugins that were not updated to use the recommended APIs.

Reviewer 2: were not --> have not been

Major changes regarding plugins include a proper immutable plugin model together with a revised plugin API.
The updated API provides hints as preparation for Maven 4.
You can enable them by passing the following argument to your build: `-Dmaven.plugin.validation=verbose`.
You should also only rely on the official Maven BOMs when developing plugins.

> If a plugin still relies on long-deprecated and now removed Plexus dependency resolution, it will no longer work.

Reviewer 4: ```suggestion
If a plugin still relies on long-deprecated and now removed Plexus dependency injection, it will no longer work.
```

Author: Updated

> If a plugin still relies on long-deprecated and now removed Plexus dependency resolution, it will no longer work

Reviewer 2: I'm not sure this is true. If it is, elaboration is needed.

Author: > I'm not sure this is true. If it is, elaboration is needed.

@cstamas @gnodet Can you give us some insight please?

and needs to be updated to use JSR-330 - see [Maven & JSR-330][26] for further information.

> **Advice**: If you are maintaining a Maven plugin, you should test it with Maven 3.9.x, pay close attention to upcoming

Reviewer 2: run-on sentence

> "With the Maven build schema preserved in amber, we can’t evolve much: we’ll stay forever with Maven 3 minor releases,
>
>> unable to implement improvements that we imagine will require seriously updating the POM schema…"
>> &mdash; <cite>[Hervé Boutemy (in Javaadvent 2021)][1]</cite>

In order for Maven to evolve, it's necessary to separate the information needed for the build from the information
needed by the consumers, but without breaking the ecosystem.
Maven 4 prepares for this and more.

This article presents and explains major changes brought by Maven 4, grouped into several topics.

## Required Java version

Maven 4 requires Java 17.
This allows Maven (and its maintainers) to make use of newer language features and improvements brought by the
JDK.

**Important note**: Java 17 is only needed to **run Maven**!
You can still compile against older Java versions using the same [compiler plugin configuration][6] as before.
If you need to compile against or use another JDK, see the [Guide to Using Toolchains][7] (or the
article [Introduction to Maven Toolchains][8] by Maven maintainer Maarten Mulders).

## POM Changes

### Build POM and consumer POM

> In Maven 3 both the build and consumers use the same POM file with model version 4.0.0.

Reviewer 2: It occurs to that the real difference is not that there are two POMs. It's that there's a new, separate consumer POM. That is, the key point is that Maven 4 generates a stripped down consumer POM that includes just the information needed by dependents, and this is what's deployed to repositories. (Or do we still deploy both and just use one/)

In other words, rename the section Consumer POM and start with , "Maven 4 generates a stripped down consumer POM that removes build information not needed by consumers,a and deploys this to the remote repository. It does not deploy the pom.xml used to build the project."

There's no need to even mention build POM here. It's much the same as in Maven 3 aside from changes listed in subsequent sections.

However, several parts of the POM are only necessary for the build while others, like the dependencies, are also
needed by the consumers.
Maven 4 therefore differentiates between a build POM and a consumer POM.
As the names suggest, the build POM contains all information needed to build the artifact, for example applied plugins
and their configuration, while the consumer POM, which is created during the Maven build, only contains what is
necessary to use an artifact as a dependency.

**Note**: See below for a comparison of the content of both POMs.

### Model version 4.1.0

> Maven 4 introduces a new POM version 4.1.0.

Reviewer 2: "Maven 4 updates the POM version to 4.1.0."

> With two types of POM, Maven 4 can make additions to the build POM as it is only used by Maven.

Reviewer 2: delete this sentence.

Version 4.1.0 adds some new elements and attributes, while others are marked as deprecated.
To not break the ecosystem, this version is only available for the build POM, while the consumer POM will still use
version 4.0.0.
Maven generates the consumer POM during the build from the build POM.

**Note**: Maven 4 will continue to build your model version 4.0.0 project.
There is no need to update your POMs to 4.1.0 if you don't need to use the new features.

### Modules are now subprojects

A Maven project is any Java project built with Maven.

> The [root directory][32] of a Maven project contains at least the source folder `src` and the project's POM file

Reviewer 2: That page doesn't seem to use or define the term "root directory"

`pom.xml`.
A Maven project may have other Maven projects within subdirectories, each with its own `pom.xml` file.
Each project within a subdirectory is called a "subproject".

> Example: A project A contains a subdirectory which contains its own POM file for project B.

Reviewer 2: A project A --> Project A

We say that B is a subproject of A.

However, Maven does not build subprojects unless they are linked from the root POM file.
In the past this was done by listing each subproject in the `<modules>` section of the parent's POM.

> Projects which have at least one subproject are called "multi-module projects", while those which have any subprojects

Reviewer 2: which have any --> which do not have any

If a plugin still relies on long-deprecated and now removed Plexus dependency injection, it will no longer work.
It needs to be updated to use JSR-330 - see [Maven & JSR-330][26] for further information.

**Advice**: If you are maintaining a Maven plugin, you should test it with Maven 3.9.x.
Pay close attention to upcoming warnings and update the plugin accordingly.

### Improved encryption

Security is important, and storing unencrypted passwords is bad practice.
Maven 3's password encryption had several serious issues and could more accurately be called "password obfuscation".
Maven 4 instead has a completely redone encryption system, based on Maven Encryption (`mvnenc`) - a standalone CLI tool.
As of now, it provides functionally equivalent operations as Maven 3 (see ["Maven: Password Encryption"][30]) but adds
improvements like a `decrypt` functionality.
A broad overview of the problems in Maven 3 and the solution in Maven 4 can be found in the
article ["Handling sensitive data in Maven"][31] by Maven maintainer Tamás Cservenák.

### Maven Resolver

The [Maven Artifact Resolver][28] is a library for working with artifact repositories and dependency resolution.

> Maven 4 will include the new 2.0 release of this library, which contains more than 150 fixes, upgrades, and

Reviewer 2: will include --> includes

<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

<head>
   <title>What's new in Maven 4?</title>
   <meta name="author" content="Matthias Bünger" />
</head>

> # What's new in Maven 4?

Reviewer 5: Inference of `<modelVersion>` is another new feature in Maven v4. You may want to add that.

Doing

```xml
<project xmlns="http://maven.apache.org/POM/4.1.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.1.0 https://maven.apache.org/xsd/maven-4.1.0.xsd">
  <modelVersion>4.1.0</modelVersion>
</project>
```

is no longer necessary: you can leave out the `<modelVersion>`  element and it will be inferred from the XML ns.

```xml
<project xmlns="http://maven.apache.org/POM/4.1.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.1.0 https://maven.apache.org/xsd/maven-4.1.0.xsd">
</project>
```

Reviewer: that feature has been challenged by https://issues.apache.org/jira/browse/MNG-8537

Author: > that feature has been challenged by https://issues.apache.org/jira/browse/MNG-8537

I'm aware of this issue and as far as I understand the discussion it's also the cause why this PR does not move forward. Maybe I should even close it completely. It's open for about 6 weeks now, some don't want to review it cause if its size and as long as it's open from time and time someone sneaks in to request documentation which should be in the explicit pages but not in a summary (@lbruun I don't mean you by this, your ping about the not necessary `<modelVersion>` is totally fine and I would instantly add it, if MNG-8537 would not exists/be closed).

This allows Maven (and its maintainers) to make use of newer language features and improvements brought by the
JDK.

**Important note**: Java 17 is only needed to **run Maven**!
You can still compile against older Java versions using the same [compiler plugin configuration][6] as before.
If you need to compile against or use another JDK, see the [Guide to Using Toolchains][7] (or the
article [Introduction to Maven Toolchains][8] by Maven maintainer Maarten Mulders).

## POM Changes

### Build POM and consumer POM

In Maven 3 both the build and consumers use the same POM file with model version 4.0.0.
However, several parts of the POM are only necessary for the build while others, like the dependencies, are also
needed by the consumers.
Maven 4 therefore differentiates between a build POM and a consumer POM.
As the names suggest, the build POM contains all information needed to build the artifact, for example applied plugins
and their configuration, while the consumer POM, which is created during the Maven build, only contains what is
necessary to use an artifact as a dependency.
This POM will only keep what is really needed to use the artifact, for example dependency information.

> **Note**: See below for a comparison of the content of both POMs.

Author: Would like to keep it, because there are some paragraphs in between this section and the actual comparison.

> **Note**: See below for a comparison of the content of both POMs.

Reviewer 2: seriously, don't. It doesn't help the reader. No one is reading this and gets here and thinks, "Gee, I wonder where a comparison of the content of both POMs is." This is a long doc and we need to take every opportunity to make it shorter and more concise.

of the root folder in your POM files, for example when you want to define the path to a `checkstyle-suppressions.xml`
file for the checkstyle plugin.
Maven 4 now provides official properties to reference the root directory in your POM configuration.
The following table shows the official properties.

| Property                   |  Scope  | Definition                                                              | Always |
|:---------------------------|:-------:|:------------------------------------------------------------------------|:------:|
| `${project.rootDirectory}` | Project | `.mvn` folder or `root` attribute in pom                                |   No   |
| `${session.topDirectory}`  | Session | Current directory or `--file` argument                                  |  Yes   |
| `${session.rootDirectory}` | Session | `.mvn` folder or `root` attribute in pom for the `topDirectory` project |   No   |

As you can see, these properties differentiate by their scope, where `project` is always related to the Maven project's
definition (you could interpret this as the POM files) and `session` is the actual execution of a Maven build and is the

> current working directory.

Reviewer 3: AFAIK, "session" is the execution of Maven, not the current working directory.

**Recommendation: Do not use `mvn clean install`.
Instead, use `mvn verify` for your regular builds!**

To improve usability when resuming a failed build, you can now use `--resume` or its short parameter `-r` to resume a
build from the subproject that last failed.
So you don't have to manually pass the name of the failed subproject as the starting point to resume from.
The reactor is now also aware of successfully built subprojects when the overall build failed and will skip rebuilding
those if you resume the build.  
With Maven 4, it's also aware of subfolder builds [MNG-6118][12], which becomes pretty handy when you only want to
execute tools (for example Jetty) on/with certain subprojects, but not on every subproject.
See Maven maintainer Maarten Mulders's article ["What's New in Maven 4" (2020)][13] for a small example.

### Further improvements

Further improvements to subprojects will also improve daily work with them.
Thanks to [MNG-6754][14], all subprojects will now have consistent timestamps in their packaged archives, while in Maven
3, each subproject had a different one.
This should make it easier to identify the archives that belong together.

> When using Maven 3, deploying a project with multiple subprojects could end up in a situation where successfully

Reviewer 5: ### Regarding "atomic deploy"
Explain how this relates to the `deployAtEnd` property ([LINK](https://maven.apache.org/plugins/maven-deploy-plugin/deploy-mojo.html)) which exists in Maven 3 too.

So this was also possible with Maven 3, right?  Is it to be understood that in Maven v4 it is no longer necessary to explicitly set `-DdeployAtEnd=true` as we do with Maven v3 ?

Author: TLDR: I removed it, because nothing has changed.

Long story: I was assuming this too, because I was sure that I did hear it multiple times in a talk/chat that this would be the case. But the truth is, that the default value for `deployAtEnd` is still `false` in Maven 4 branch of the deploy plugin (same for install plugin). It "just" got several improvements, but even those slightly before the first maven 4 release, so also available in Maven 3 version of the plugin. Thanks for pointing this out!

artifact.
These dependent projects are called the "consumers".

This made Maven more than a tool; it became a whole ecosystem with many dependencies on the POM, especially the
Maven Central repository, other build tools, and IDEs.
Thus, any change in the POM's schema forces each participant of the ecosystem to either adopt the change or drop
support.
Thus, the Maven POM syntax became fixed, unable to change.

> "With the Maven build schema preserved in amber, we can’t evolve much: we’ll stay forever with Maven 3 minor releases,
>
>> unable to implement improvements that we imagine will require seriously updating the POM schema…"
>> &mdash; <cite>[Hervé Boutemy (in Javaadvent 2021)][1]</cite>

In order for Maven to evolve, it's necessary to separate the information needed for the build from the information
needed by the consumers, but without breaking the ecosystem.
Maven 4 prepares for this and more.

This article presents and explains major changes brought by Maven 4, grouped into several topics.

## Required Java version

Maven 4 requires Java 17.
This allows Maven (and its maintainers) to make use of newer language features and improvements brought by the

> JDK.

Reviewer 6: rather the JRE (JDK is only optionally used with m-compiler-p).

Author: Changed to end the sentence after "improvements".

You can now use variables as versions in your configuration.

Example

```xml

<groupId>my.groupId</groupId>
<artifactId>my.artifactId</artifactId>
<version>${revision}</version>
```

You have to provide a value for this variable when starting the build, for example by using a `maven.config` file.
In CI pipelines it's commonly done using a parameter, for example `mvn verify -Drevision=4.0.1`.

Maven maintainer Karl Heinz Marbaise shows a larger example in
his [article "Maven 4 - Part I - Easier Versions" (2024)][21].

### Reactor improvements and fixes

Building a project with multiple subprojects can cause confusion in Maven 3.
When the build of subproject B, which requires subproject A, is failing for whatever reason, Maven 3 is telling
the user to fix the error and then resume the build with `--resume-from :<nameOfTheFailingSubproject>`.
But doing this instantly fails the build again, because the required subproject A couldn't be found (as it was not
rebuilt).
Using `--also-make :<nameOfTheDependentSubproject>` was no help in the past as it was ignored due to the long-standing
bug [MNG-6863][11] - which is finally fixed with Maven 4!

> **Recommendation: Do not use `mvn clean install`.

Reviewer 7: I think we need to write explicitly somewhere
that for the building a subproject we don't need to `install` the root project.

Author: > I think we need to write explicitly somewhere that for the building a subproject we don't need to `install` the root project.

I think https://maven.apache.org/guides/mini/guide-multiple-subprojects-4.html is the correct place for this.

Reviewer 2: agreed, this document is just for a quick summary of what's new. It's not for documenting the new or existing features in detail.

## Maven plugins, security, and tools

### Maven plugins

As mentioned above, Maven 4 contains significant code and API updates, resulting in breaking changes for (very) old
Maven plugins that have not been updated to use the recommended APIs.
Major changes regarding plugins include a proper immutable plugin model together with a revised plugin API.
The updated API provides hints as preparation for Maven 4.
You can enable them by passing the following argument to your build: `-Dmaven.plugin.validation=verbose`.
You should also only rely on the official Maven BOMs when developing plugins.
If a plugin still relies on long-deprecated and now removed Plexus dependency injection, it will no longer work.
It needs to be updated to use JSR-330 - see [Maven & JSR-330][26] for further information.

**Advice**: If you are maintaining a Maven plugin, you should test it with Maven 3.9.x.
Pay close attention to upcoming warnings and update the plugin accordingly.

### Improved encryption

Security is important, and storing unencrypted passwords is bad practice.
Maven 3's password encryption had several serious issues and could more accurately be called "password obfuscation".

> Maven 4 instead has a completely redone encryption system, based on Maven Encryption (`mvnenc`) - a standalone CLI tool.

Reviewer 6: mvnenc is just a frontend, the underlying backend is rather https://github.com/codehaus-plexus/plexus-sec-dispatcher. Compare with #654.

Reviewer 4: I would not go into this distinction: plexus-sec-dispatcher is NOT an API, is not something people should use or reuse or even made aware. We already had problems with it in Maven3.

Reviewer 6: But `based on Maven Encryption (`mvnenc`) - a standalone CLI tool.` feels wrong to me. I would rather remove the sentence then altogether.

Reviewer 6: So what API is proposed by Maven for sensitive values e.g. for Mojos or extensions?

Reviewer 2: "plexus-sec-dispatcher is NOT an API, is not something people should use or reuse or even made aware. " -- Hyrum's Law rides again

Author: > But
> 
> > based on Maven Encryption (`mvnenc`) - a standalone CLI tool.
> 
> feels wrong to me. I would rather remove the sentence then altogether.

The phrasing is based on @cstamas blog about it:

> Maven4 password encryption is handled by the new CLI tool: mvnenc.

As of now, it provides functionally equivalent operations as Maven 3 (see ["Maven: Password Encryption"][30]) but adds
improvements like a `decrypt` functionality.
A broad overview of the problems in Maven 3 and the solution in Maven 4 can be found in the
article ["Handling sensitive data in Maven"][31] by Maven maintainer Tamás Cservenák.

### Maven Resolver

The [Maven Artifact Resolver][28] is a library for working with artifact repositories and dependency resolution.
Maven 4 includes the new 2.0 release of this library, which contains more than 150 fixes, upgrades, and
improvements, for example, a Java native HTTP-Client - thanks to raising the Java version requirement to JDK 17!
Another major difference compared to Maven 3 is that in the new major version, the resolver is hidden behind the new

> Maven API and is no longer used directly by plugins.

Reviewer 8: This might be a good place to document the property `-Dmaven.resolver.dependencyManagerTransitivity=false`

[MNG-8333 Document Maven4 transitive depMgt handling](https://issues.apache.org/jira/browse/MNG-8333)

Author: > This might be a good place to document the property `-Dmaven.resolver.dependencyManagerTransitivity=false`
> 
> [MNG-8333 Document Maven4 transitive depMgt handling](https://issues.apache.org/jira/browse/MNG-8333)

I was thinking about this quite a time and started to change the text here, but in the end I don't want to add it, because this argument should be in a own Maven resolver guide, which also describes WHAT transitive dependency management is (including examples).

Reviewer 2: I agree. This is a what's new page. It should not be the only or primary source of documentation for any feature. That belongs elsewhere.

<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

<head>
   <title>What's new in Maven 4?</title>
   <meta name="author" content="Matthias Bünger" />
</head>

# What's new in Maven 4?

Maven is over 20 years old, and is one of the most used build tools in the Java world.
Throughout the years, one important rule has been maintaining the highest backward compatibility possible, especially of
its [POM file][2] with Model version 4.0.0.

The POM file fulfills two elementary needs.

> On the one side, the POM holds all the information and configuration, which are only needed to build the artifact.

Reviewer 2: On the on side --> First

> After the artifact was created, these build information are not relevant anymore.

Reviewer 2: After the artifact is created, this build information is not relevant anymore.

> On the other side, the POM also contains information, e.g. dependencies, which are needed by projects, which want to use

Reviewer 2: On then other side --> Second
no comma after projects

> the artifact, e.g. dependencies.

Reviewer 2: delete ", e.g. dependencies" as it confuses the dependent with the dependencies

Author: Oh it's twice. Shall I delete both or only the second on line 35 where your comment it?

> These dependent projects, which "consume" an artifact (and its POM), are called the "consumers" (of an artifact).

Reviewer 2: delete "which "consume" an artifact (and its POM),"

This made the Maven more than a tool; it became a whole ecosystem with many dependencies on the POM, especially the
Maven Central repository, other build tools, and IDEs.

> This results in the situation, that any change in the POM's schema forces each participant of the ecosystem to either

Reviewer 2: situation, that --> situation that

adopt the change or drop support.

> So Maven's POM file got cemented, not able to change.

Reviewer 2: Thus the Maven POM syntax became fixed, unable to change.

> "With the Maven build schema preserved in amber, we can’t evolve much: we’ll stay forever with Maven 3 minor releases,
> unable to implement improvements that we imagine will require seriously updating the POM schema…"
> &mdash; <cite>[Hervé Boutemy (in Javaadvent 2021)][1]</cite>

But Maven should be able to advance.

> For this, one important thing that's needed is to separate the information needed for the build from those needed by the

Reviewer 2: those --> the information

> consumers, but without breaking in the ecosystem.

Reviewer 2: in the --> the

> Maven 4 will prepare for this and more.

Reviewer 2: will prepare --> prepares

This article presents and explains major changes brought by Maven 4, grouped into several topics.

## POM Changes

### Build-POM and Consumer-POM

> As written in the introduction, Model version 4.0.0 is used not only by the build but also by consumers of the

Reviewer 2: Delete "As written in the introduction, "

It's important to note that model version 4.0.0 is what's used by Maven 3 and model version 4.1 is used by Maven 4 since this is surprising.

Author: Will think about another wording tomorrow. Need to go to local JUG now.

artifact.
However, several parts of the POM are only necessary for the build while others, like the dependencies, are also
needed by the consumers.

> Maven 4 will therefore differentiate between a "Build-POM" and a "Consumer-POM".

Reviewer 2: will therefore differentiate --> therefore differentiates

> As the names suggest, the "Build-POM" will contain all information needed to build the artifact, e.g., applied plugins

Reviewer 2: will contain --> contains

It's likely this doc exists in the same time frame as Maven 4 so use present tense when describing it

and their configuration, while the "Consumer-POM", which is created during the Maven build, only contains what is

> necessary.

Reviewer 2: necassary for ????

  </parent>

  <artifactId>myOwnSubprojectArtifactId</artifactId>

  <dependencies>
    <dependency>
      <groupId>the.dependent.subproject.groupId</groupId>
      <artifactId>the.dependent.subproject.artifactId</artifactId>
    </dependency>
  </dependencies>
</project>
```

### Full support of CI-friendly variables

Maven 3.5.0 introduced partial support for CI-friendly variables, e.g., `${revision}`, in your POM files.
However, this still required the usage of the [Flatten Maven Plugin][20] for full functionality.
Since Maven 4, no additional plugin is needed; full built-in support is provided.
You can now use variables as versions in your configuration, e.g.,

```xml

<groupId>my.groupId</groupId>
<artifactId>my.artifactId</artifactId>
<version>${revision}</version>
```

You have to provide a value for this variable when starting the build, for example by using a `maven.config` file.
In CI pipelines it's commonly done using a parameter, e.g., `mvn verify -Drevision=4.0.1`.

Maven maintainer Karl Heinz Marbaise shows a larger example in
his [article "Maven 4 - Part I - Easier Versions" (2024)][21].

### Reactor improvements and fixes

> Building a project with multiple subprojects could cause trouble when one subproject was dependent on one of the

Reviewer 2: was dependent --> depended

others and its own build failed for whatever reason.

> Maven was telling the user to (fix the error and then) resume the build with

Reviewer 2: no parentheses

`--resume-from :<nameOfTheFailingSubproject>`, which instantly fails the build again as the needed other subproject
couldn't be found (as it was not rebuilt).
Using `--also-make :<nameOfTheDependentSubproject>` was no help in the past as it was ignored due to the long-standing
bug [MNG-6863][11] - which is finally fixed with Maven 4!

> **Recommendation: Do not use `mvn clean install`, but `mvn verify` for your regular builds!**

Reviewer 2: , but --> . Instead

#### Lifecycle changed from graph to tree

> Up to Maven 3 the lifecycle was a literally an ordered list which contains all phases.

Reviewer: `Up to Maven 3, the lifecycle was literally an ordered list containing all phases.`

Author: Thanks again!

### Maven Resolver

The [Maven Artifact Resolver][28] is a library for working with artifact repositories and dependency resolution.
Maven 4 will include the new 2.0 release of this library, which contains more than 150 fixes, upgrades, and
improvements, e.g., a Java native HTTP-Client - thanks to raising the Java version requirement to JDK 17!
Another big difference in comparison with Maven 3 is that with the new major version the resolver aims to be hidden

> behind the new Maven API and thus not used directly anymore.

Reviewer: `Another major difference compared to Maven 3 is that in the new major version, the resolver is hidden behind the new Maven API and is no longer used directly by plugins.`

Author: Thanks again!

Another major difference compared to Maven 3 is that in the new major version, the resolver is hidden behind the new
Maven API and is no longer used directly by plugins.

### Maven Shell

When you run any `mvn` command, the full chain of booting Java, Maven, load config, do the work, tear down and exit the
process is done - **every single time**.
To improve performance and reduce build time, you can use the [Maven Daemon][29] (`mvnd`), which manages a pool of N
resident Maven processes.
With Maven 4 you can also use the brightly polished "Maven Shell" (`mvnsh`), which will keep exactly one resident Maven

> process as long as the shell is opened.

Reviewer: ```
### Maven Shell

Each time you run an `mvn` command, the entire process chain is executed: booting Java, starting Maven, loading the configuration, performing the task, tearing down, and exiting — **every single time**.  

To improve performance and reduce build times, you can use the [Maven Daemon][29] (`mvnd`), which manages a pool of resident Maven processes.  

With Maven 4, you can also take advantage of the newly defined "Maven Shell" (`mvnsh`), which keeps a single Maven process running for as long as the shell remains open.
```

Minor improvements to this section.
